### DISCUSSION TITLE
Tasmota-embedded MQTT broker

### DISCUSSION THREAD

**Discussion by bam80** [arendst/Tasmota#22711] [Q&A] [answered]
Hi,
the question is about situations when you can't have a separate MQTT broker device.

Is there any way to embed it to Tasmota itself then?
If no, any better solutions/protocols maybe?

PS:
Use case:
provide fallback mean for Web-application to connect to Tasmota directly if remote MQTT broker is not accessible (and local one can't be set up for some reasons).

**Comment by Jason2866** [comment]
No embbeded broker. esp32 has not enough resources for this.
mqtt is lightweight and the most used protocol for IoT devices
The "m" in Tasmota stands for mqtt. It is designed for using mqtt.
So no there is no better protocol for Tasmota than mqtt.

**Reply by bam80** [reply]
Thanks.
So how do we handle the situation in question then?

**Reply by joba-1** [reply]
Always depends on the use case but lacking that info: second best is sending http requests.

**Reply by sfromis** [reply]
One of the major drawbacks of using http is not getting updates or (in default setup) command responses. Can usually work for "fire and forget" commands.

**Reply by bam80** [reply]
While I see there is no "good enough" solution, at least it's something.
Thanks to all.

**Reply by bam80** [reply]
> No embbeded broker. esp32 has not enough resources for this.

It seems as week argument as competing firmware does support this:
https://github.com/Mit4el/IoTManager/tree/ver4dev/src%2Fmodules%2Fexec%2FBrokerMQTT

**Reply by s-hadinger** [reply, marked as answer]
Anyways Tasmota does not support acting as an mqtt broker. If you want a good reason : nobody invested time to make it happen.

**Comment by HofeBY2** [comment]
@bam80
It is possible to send command´s between Tasmota devices without an mqtty brooker.
search for KNX over TCP

**Reply by barbudor** [reply]
I don't know if there is a variant of KNX over TCP but for sure Tasmota implementation use UDP Multicast, not TCP

**Reply by sfromis** [reply]
Sending commands between Tasmota devices can work via http (using WebQuery/WebSend), or via device groups. Still, MQTT is the preferred strategy.

**Comment by bam80** [comment]
I added my use case to the description, if it's helpful.

Thanks to all for your responses and happy upcoming holidays!

**Reply by sfromis** [reply]
If you intend to expose Tasmota devices to the big bad Internet, don't create such a major security hole, without a strong security layer in between.

**Reply by bam80** [reply]
The idea is in case of internet outage or any other reason of MQTT broker inaccessibility, you still able to use your Web app in your house via local Wi-Fi by "redirecting" it to Tasmota device directly.

**Reply by sfromis** [reply]
No web app, but as already mentioned, you *can* use a browser for one device at a time. Thus, you can bookmark each on a phone etc home screen (like in a folder) to have relatively quick access.

**Reply by bam80** [reply]
Tasmota's Web interface is rudimentary and in-acceptable for our needs. I'm designing custom Web app (with MQTT backend) for my project.

**Reply by joba-1** [reply]
I still don’t quite understand the usecase. You can host a web app, but not an mqtt broker?
The broker is out there on the internet and not with your tasmota devices? 
How would a switch to a local backup mqtt broker work? 
Doing this coordinated for all clients so they can still all communicate with each other is a complicated task and usually requires some cluster management software which would be hosted where?

**Reply by bam80** [reply]
> You can host a web app, but not an mqtt broker?

It's PWA Web app. Once loaded, it doesn't need hosting any more.

> The broker is out there on the internet and not with your tasmota devices?

For example.

> How would a switch to a local backup mqtt broker work?

Supposedly, the App would support that.

All of that is not hard requirement for my project, as we can always use the alternative firmware supporting this.
Just wondering..

**Comment by wjohn007** [comment]
> Use case:
> provide fallback mean for Web-application to connect to Tasmota directly if remote MQTT broker is not accessible (and local one can't be set up for some reasons).

I had similar requirements for my project. Sending fire-and-forget messages via UDP broadcast turned out to be a practical solution.
The implementation was done in Berry and is modeled on the standard Berry-API for MQTT.

The code-example on the recipient side
```berry
def UdpMessageHandler(topic,payload)
  print("UdpMessageHandler received topic:",topic," payload:",payload)
end

udpBroker = UdpBroker("broker")
udpBroker.subscribe("MyCoolTopic",UdpMessageHandler)
```
The code-example of the sender side

```berry
udpBroker = UdpBroker("broker")
udpBroker.publish("MyCoolTopic","Hello world")
```

This means that every Tasmota controller can subscribe to a topic and receive the desired messages. 
A physical broker is not necessary here. Everything can be implemented with Tasmota tools.

**Reply by sfromis** [reply]
The above examples are based on https://github.com/wjohn007/Berry-UdpBroker/

Would take more code if the intention is to get commands across, basically to emulate the existing `WebQuery` / `WebSend` commands.
